<launch>
  <!-- Set this to your camera's name -->
  <arg name="primary_cam" value="jetson_cam" />
  <arg name="primary_cam_info" value="file:///home/nvidia/.camera_info/jetson_cam_1280x720.yaml" />
  <arg name="secondary_cam" value="logitech_cam" />
  <arg name="secondary_cam_info" value="file:///home/nvidia/.camera_info/logitech_cam_640x360.yaml" />

  <arg name="gui" default="true"/>
  <arg name="feed" default="true"/>
  
  <!-- Starting team colour -->
  <arg name="side" default="orange"/>



  <!-- Camera driver for onboard Jetson camera -->
  <node pkg="gscam" type="gscam" name="$(arg primary_cam)_driver_node">
    <param name="camera_name" value="$(arg primary_cam)" />
    <param name="camera_info_url" value="$(arg primary_cam_info)" />
    <param name="frame_id" value="primary_camera" />
    <param name="gscam_config" value="nvcamerasrc ! video/x-raw(memory:NVMM), width=(int)1280, height=(int)720,format=(string)I420, framerate=(fraction)20/1 ! nvvidconv flip-method=0 ! video/x-raw, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR" />
    
    <remap from="camera/image_raw" to="$(arg primary_cam)/image_raw" />
    <remap from="set_camera_info" to="$(arg primary_cam)/set_camera_info" />
  </node>

  <!-- Camera driver for secondary Logitech camera -->
  <node pkg="cv_camera" type="cv_camera_node" name="$(arg secondary_cam)_driver_node">
      <param name="device_path" value="/dev/video1" />
      <param name="frame_id" value="secondary_camera" />
      <param name="image_width" value="640" />
      <param name="image_height" value="360" />
      <param name="camera_info_url" value="$(arg secondary_cam_info)" />

      <remap from="$(arg secondary_cam)_driver_node/image_raw" to="$(arg secondary_cam)/image_raw" />
  </node>



  <!-- Display raw & recitfied image feeds from camera drivers -->
  <node ns="$(arg primary_cam)" pkg="image_proc" type="image_proc" name="image_rectifier_node" />  
  <node ns="$(arg secondary_cam)" pkg="image_proc" type="image_proc" name="image_rectifier_node" />                                 

  <!-- <node pkg="image_view" type="image_view" name="primary_raw_feed" >   --> 
  <!--   <remap from="image" to="/$(arg primary_cam)/image_raw" />          --> 
  <!-- </node>                                                              --> 

  <!-- <node pkg="image_view" type="image_view" name="secondary_raw_feed" > --> 
  <!--   <remap from="image" to="/$(arg secondary_cam)/image_raw" />        --> 
  <!-- </node>                                                              --> 
              
  <!-- <node ns="$(arg primary_cam)" pkg="image_view" type="image_view" name="image_rect_display" >    --> 
  <!--   <remap from="image" to="/$(arg primary_cam)/image_rect_color" />                              --> 
  <!-- </node>                                                                                         --> 

  <!-- <node ns="$(arg secondary_cam)" pkg="image_view" type="image_view" name="image_rect_display" >  -->   
  <!--   <remap from="image" to="/$(arg secondary_cam)/image_rect_color" />                            --> 
  <!-- </node>                                                                                         --> 



  <!-- Apriltags2_ros for multiple cameras -->
  <arg name="launch_prefix" default="" /> 
  <arg name="primary_node_namespace" default="$(arg primary_cam)/apriltags_node" />
  <arg name="primary_camera_frame" default="primary_cam" />
  <arg name="secondary_node_namespace" default="$(arg secondary_cam)/apriltags_node" />
  <arg name="secondary_camera_frame" default="secondary_cam" />

  <rosparam command="load" file="$(find eurobot_localisation)/config/settings.yaml" ns="$(arg primary_node_namespace)" />
  <rosparam command="load" file="$(find eurobot_localisation)/config/tags.yaml"     ns="$(arg primary_node_namespace)" />
  <rosparam command="load" file="$(find eurobot_localisation)/config/settings.yaml" ns="$(arg secondary_node_namespace)" />
  <rosparam command="load" file="$(find eurobot_localisation)/config/tags.yaml"     ns="$(arg secondary_node_namespace)" />
  
  <node ns="$(arg primary_cam)" pkg="apriltags2_ros" type="apriltags2_ros_continuous_node" name="apriltags_node" clear_params="true" output="screen" launch-prefix="$(arg launch_prefix)" >
    <param name="camera_frame" type="str" value="$(arg primary_camera_frame)" />
    <param name="publish_tag_detections_image" type="bool" value="true" />      
  </node>
  
  <node ns="$(arg secondary_cam)" pkg="apriltags2_ros" type="apriltags2_ros_continuous_node" name="apriltags_node" clear_params="true" output="screen" launch-prefix="$(arg launch_prefix)" >
    <param name="camera_frame" type="str" value="$(arg secondary_camera_frame)" />
    <param name="publish_tag_detections_image" type="bool" value="true" />      
  </node>



  <!-- Display tags detected from both cameras -->
  <node if="$(arg feed)" ns="$(arg primary_cam)" pkg="image_view" type="image_view" name="tag_detection_feed" >     
    <remap from="image" to="/$(arg primary_cam)/tag_detections_image" />       
  </node>        
  
  <node if="$(arg feed)" ns="$(arg secondary_cam)" pkg="image_view" type="image_view" name="tag_detection_feed" >     
    <remap from="image" to="/$(arg secondary_cam)/tag_detections_image" />       
  </node>        


    
  <!-- Run localisation module for each camera feed -->
  <node ns="$(arg primary_cam)" pkg="eurobot_localisation" type="localisation_module" name="localisation_node" >
    <param name="name"   type="str"    value="$(arg primary_cam)" />
    <param name="colour" type="str"    value="$(arg side)" />

    <!-- values for camera pose relative to robot base -->
    <param name="rx"     type="double" value="0.05" />
    <param name="ry"     type="double" value="0.00" />
    <param name="rz"     type="double" value="0.15" />
    
    <param name="roll"   type="double" value="-1.57" />
    <param name="pitch"  type="double" value="0.00" />
    <param name="yaw"    type="double" value="-1.57" />
  </node>

  <node ns="$(arg secondary_cam)" pkg="eurobot_localisation" type="localisation_module" name="localisation_node" >
    <param name="name"   type="str"    value="$(arg secondary_cam)" />
    <param name="colour" type="str"    value="$(arg side)" />

    <!-- values for camera pose relative to robot base -->
    <param name="rx"     type="double" value="0.05" />
    <param name="ry"     type="double" value="0.00" />
    <param name="rz"     type="double" value="0.30" />
    
    <param name="roll"   type="double" value="-1.57" />
    <param name="pitch"  type="double" value="0.00" />
    <param name="yaw"    type="double" value="-1.57" />
  </node>



  <!-- Display on Rviz for testing purposes -->
  <node if="$(arg gui)" pkg="eurobot_localisation" type="eurobot_tf_broadcaster_node" name="eurobot_arena_tf_generator" >
    <param name="colour" type="str" value="$(arg side)" />
  </node>

  <node if="$(arg gui)" pkg="rviz" type="rviz" name="tf_visualiser" output="screen" args="-d $(find eurobot_localisation)/config/eurobot_localisation.rviz"/>


</launch>
